### ISSUE TITLE
Implementing PPG (Phasic Policy Gradient)

### ISSUE THREAD

**Issue by vwxyzjn** [vwxyzjn/cleanrl#32]
## Problem Description

PPG (Phasic Policy Gradient, https://arxiv.org/abs/2009.04416) seems to be a major update to PPO that improves the sample-efficiency of PPO. PPG is examined on Procgen (#29) in the paper, which we should replicate in the [Open RL Benchmark](http://benchmark.cleanrl.dev/). In addition, it might be intersting to examine its performance in Atari and continuous control tasks as well.

**Comment by vwxyzjn** [comment]
PPG has pretty large GPU memory requirement due to the its use of batch to store obs. To make it easier to use smaller GPU memory, we experiment with gradient accumulation.

Below are the gradient accumulation results for the value loss

```python
In [74]: optimizer.zero_grad()
    ...: args.aux_batch_size = int(args.batch_size * args.n_iteration)
    ...: args.aux_minibatch_size  = int(args.aux_batch_size // (args.n_aux_minibatch))
    ...: aux_inds = np.arange(args.aux_batch_size,)
    ...: print("aux phase starts")
    ...: for i, start in enumerate(range(0, args.aux_batch_size, args.aux_minibatch_size)):
    ...:     end = start + args.aux_minibatch_size
    ...:     aux_minibatch_ind = aux_inds[start:end]
    ...:     m_aux_obs = aux_obs[aux_minibatch_ind].to(device)
    ...:     m_aux_returns = aux_returns[aux_minibatch_ind].to(device)
    ...: 
    ...:     new_values = agent.get_value(m_aux_obs).view(-1)
    ...:     new_aux_values = agent.get_aux_value(m_aux_obs).view(-1)
    ...:     kl_loss = td.kl_divergence(agent.get_pi(m_aux_obs), old_agent.get_pi(m_aux_obs)).mean()
    ...: 
    ...:     real_value_loss = 0.5 * ((new_values - m_aux_returns) ** 2).mean()
    ...:     aux_value_loss = 0.5 * ((new_aux_values - m_aux_returns) ** 2).mean()
    ...:     joint_loss = aux_value_loss + args.beta_clone * kl_loss
    ...: 
    ...:     (real_value_loss).backward()
    ...:     # nn.utils.clip_grad_norm_(agent.parameters(), args.max_grad_norm)
    ...:     print(agent.critic[-1].weight.grad.sum())
    ...:     break
    ...:     
aux phase starts
tensor(-6.6915, device='cuda:0')

In [75]: optimizer.zero_grad()
    ...: args.aux_batch_size = int(args.batch_size * args.n_iteration)
    ...: args.aux_minibatch_size  = int(args.aux_batch_size // (args.n_aux_minibatch * args.n_aux_grad_accum))
    ...: aux_inds = np.arange(args.aux_batch_size,)
    ...: print("aux phase starts")
    ...: for i, start in enumerate(range(0, args.aux_batch_size, args.aux_minibatch_size)):
    ...:     end = start + args.aux_minibatch_size
    ...:     aux_minibatch_ind = aux_inds[start:end]
    ...:     m_aux_obs = aux_obs[aux_minibatch_ind].to(device)
    ...:     m_aux_returns = aux_returns[aux_minibatch_ind].to(device)
    ...: 
    ...:     new_values = agent.get_value(m_aux_obs).view(-1)
    ...:     new_aux_values = agent.get_aux_value(m_aux_obs).view(-1)
    ...:     kl_loss = td.kl_divergence(agent.get_pi(m_aux_obs), old_agent.get_pi(m_aux_obs)).mean()
    ...: 
    ...:     real_value_loss = 0.5 * ((new_values - m_aux_returns) ** 2).mean()
    ...:     aux_value_loss = 0.5 * ((new_aux_values - m_aux_returns) ** 2).mean()
    ...:     joint_loss = aux_value_loss + args.beta_clone * kl_loss
    ...: 
    ...:     loss = (real_value_loss) / args.n_aux_grad_accum
    ...:     loss.backward()
    ...:     # nn.utils.clip_grad_norm_(agent.parameters(), args.max_grad_norm)
    ...:     if (i+1) % args.n_aux_grad_accum == 0:
    ...:         print(agent.critic[-1].weight.grad.sum())
    ...:         break
    ...:         
aux phase starts
tensor(-6.6915, device='cuda:0')
```

kl loss:
```python
optimizer.zero_grad()
args.aux_batch_size = int(args.batch_size * args.n_iteration)
args.aux_minibatch_size  = int(args.aux_batch_size // (args.n_aux_minibatch))
aux_inds = np.arange(args.aux_batch_size,)
print("aux phase starts")
for i, start in enumerate(range(0, args.aux_batch_size, args.aux_minibatch_size)):
    end = start + args.aux_minibatch_size
    aux_minibatch_ind = aux_inds[start:end]
    m_aux_obs = aux_obs[aux_minibatch_ind].to(device)
    m_aux_returns = aux_returns[aux_minibatch_ind].to(device)

    new_values = agent.get_value(m_aux_obs).view(-1)
    new_aux_values = agent.get_aux_value(m_aux_obs).view(-1)
    kl_loss = td.kl_divergence(agent.get_pi(m_aux_obs), old_agent.get_pi(m_aux_obs)).mean()

    real_value_loss = 0.5 * ((new_values - m_aux_returns) ** 2).mean()
    aux_value_loss = 0.5 * ((new_aux_values - m_aux_returns) ** 2).mean()
    joint_loss = aux_value_loss + args.beta_clone * kl_loss

    (kl_loss).backward()
    # nn.utils.clip_grad_norm_(agent.parameters(), args.max_grad_norm)
    print(agent.actor.weight.grad.sum())
    break
    
aux phase starts
tensor(5.4805e-07, device='cuda:0')

optimizer.zero_grad()
args.aux_batch_size = int(args.batch_size * args.n_iteration)
args.aux_minibatch_size  = int(args.aux_batch_size // (args.n_aux_minibatch * args.n_aux_grad_accum))
aux_inds = np.arange(args.aux_batch_size,)
print("aux phase starts")
for i, start in enumerate(range(0, args.aux_batch_size, args.aux_minibatch_size)):
    end = start + args.aux_minibatch_size
    aux_minibatch_ind = aux_inds[start:end]
    m_aux_obs = aux_obs[aux_minibatch_ind].to(device)
    m_aux_returns = aux_returns[aux_minibatch_ind].to(device)

    new_values = agent.get_value(m_aux_obs).view(-1)
    new_aux_values = agent.get_aux_value(m_aux_obs).view(-1)
    kl_loss = td.kl_divergence(agent.get_pi(m_aux_obs), old_agent.get_pi(m_aux_obs)).mean()

    real_value_loss = 0.5 * ((new_values - m_aux_returns) ** 2).mean()
    aux_value_loss = 0.5 * ((new_aux_values - m_aux_returns) ** 2).mean()
    joint_loss = aux_value_loss + args.beta_clone * kl_loss

    loss = (kl_loss) / args.n_aux_grad_accum
    loss.backward()
    # nn.utils.clip_grad_norm_(agent.parameters(), args.max_grad_norm)
    if (i+1) % args.n_aux_grad_accum == 0:
        print(agent.actor.weight.grad.sum())
        break
        
aux phase starts
tensor(5.8156e-07, device='cuda:0')
```

So apparently gradient accumulation works with the value loss but something in `td.kl_divergence` is throwing off the calculation for the policy related losses. For the moment let's ignore it due to the benefits of using gradient accumulation.

**Comment by vwxyzjn** [comment]
I noticed some nit different between  the PPO implementation from your repos and the openai/baselines’s PPO implementation. For example, the `policy_head` and `value_head` of PPG’s PPO implementation [are both initialized with scale 0.1](https://github.com/openai/phasic-policy-gradient/blob/7295473f0185c82f9eb9c1e17a373135edd8aacc/phasic_policy_gradient/ppg.py#L99-L108), while the openai/baselines’sPPO implementation uses [scale 0.01 for the `policy_head` and `1` for the `value_head`](https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/common/policies.py#L49-L63). 

For these nit differences, some implementation decisions are no clear to me as far as how to implement PPG. For this reason, closing this issue now.
